File: scala/scalasearch/src/test/scala/scalasearch/SearchResultTest.scala
package scalasearch

import org.junit.Assert.assertEquals
import org.scalatest.funsuite.AnyFunSuite
import scalafind.FileType.FileType
import scalafind.{ConsoleColor, FileResult, FileType}

import java.nio.file.Paths

class SearchResultTest extends AnyFunSuite {

  def getSearchResult(pattern: String, path: String, fileType: FileType, lineNum: Int,
                      matchStartIdx: Int, matchEndIdx: Int, line: Option[String],
                      linesBefore: Seq[String]=Seq.empty[String],
                      linesAfter: Seq[String]=Seq.empty[String]): SearchResult = {
    val fileResult = FileResult(Paths.get(path), fileType)
    SearchResult(pattern.r, Some(fileResult), lineNum, matchStartIdx, matchEndIdx, line, linesBefore, linesAfter)
  }

  test("test single-line search result") {
    val pattern = "Search"
    val filePath = "~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"
    val lineNum = 10
    val matchStartIndex = 15
    val matchEndIndex = 23
    val line = "\tpublic class Searcher\n"
    val searchResult = getSearchResult(pattern, filePath, FileType.Code, lineNum,
      matchStartIndex, matchEndIndex, Some(line))
    assertEquals(pattern, searchResult.searchPattern.pattern.pattern())
    assertEquals(filePath, searchResult.file.get.path.toString)
    assertEquals(lineNum, searchResult.lineNum)
    assertEquals(matchStartIndex, searchResult.matchStartIndex)
    assertEquals(matchEndIndex, searchResult.matchEndIndex)
    assertEquals(line, searchResult.line.get)
    assertEquals(0, searchResult.linesBefore.length)
    assertEquals(0, searchResult.linesAfter.length)
  }

  test("test single-line search result format") {
    val settings = SearchSettings(colorize = false)
    val pattern = "Search"
    val filePath = "~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"
    val lineNum = 10
    val matchStartIndex = 15
    val matchEndIndex = 23
    val line = "\tpublic class Searcher\n"
    val searchResult = getSearchResult(pattern, filePath, FileType.Code, lineNum,
      matchStartIndex, matchEndIndex, Some(line))
    val expectedOutput = "%s: %d: [%d:%d]: %s".format(filePath,
      lineNum, matchStartIndex, matchEndIndex, line.trim)
    val output = new SearchResultFormatter(settings).format(searchResult)
    assertEquals(expectedOutput, output)
  }

  test("test single-line longer than maxlinelength search result") {
    val settings = SearchSettings(colorize = false, maxLineLength = 100)
    val pattern = "maxlen"
    val filePath = "./maxlen.txt"
    val lineNum = 1
    val matchStartIndex = 53
    val matchEndIndex = 59
    val line =
      "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789"
    val searchResult = getSearchResult(pattern, filePath, FileType.Text, lineNum,
      matchStartIndex, matchEndIndex, Some(line), List.empty, List.empty)
    val expectedLine = "...89012345678901234567890123456789012345678901maxlen89012345678901234567890123456789012345678901..."
    val expectedOutput = "%s: %d: [%d:%d]: %s".format(filePath,
      lineNum, matchStartIndex, matchEndIndex, expectedLine)
    val output = new SearchResultFormatter(settings).format(searchResult)
    assertEquals(expectedOutput, output)
  }

  test("test single-line longer colorize search result") {
    val settings = SearchSettings(colorize = true, maxLineLength = 100)
    val pattern = "maxlen"
    val filePath = "./maxlen.txt"
    val lineNum = 1
    val matchStartIndex = 53
    val matchEndIndex = 59
    val line =
      "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789"
    val searchResult = getSearchResult(pattern, filePath, FileType.Text, lineNum,
      matchStartIndex, matchEndIndex, Some(line), List.empty, List.empty)
    val expectedLine = "...89012345678901234567890123456789012345678901" +
      ConsoleColor.GREEN.toString +
      "maxlen" +
      ConsoleColor.RESET.toString +
      "89012345678901234567890123456789012345678901..."
    val expectedOutput = "%s: %d: [%d:%d]: %s".format(filePath,
      lineNum, matchStartIndex, matchEndIndex, expectedLine)
    val output = new SearchResultFormatter(settings).format(searchResult)
    assertEquals(expectedOutput, output)
  }

  test("test match length longer colorize search result") {
    val settings = SearchSettings(colorize = true, maxLineLength = 100)
    val pattern = "\\d+maxlen\\d+"
    val filePath = "./maxlen.txt"
    val lineNum = 10
    val matchStartIndex = 1
    val matchEndIndex = 110
    val line =
      "0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789"
    val searchResult = getSearchResult(pattern, filePath, FileType.Text, lineNum,
      matchStartIndex, matchEndIndex, Some(line), List.empty, List.empty)
    val expectedLine = ConsoleColor.GREEN.toString +
      "0123456789012345678901234567890123456789012345678901maxlen890123456789012345678901234567890123456" +
      ConsoleColor.RESET.toString +
      "..."
    val expectedOutput = "%s: %d: [%d:%d]: %s".format(filePath,
      lineNum, matchStartIndex, matchEndIndex, expectedLine)
    val output = new SearchResultFormatter(settings).format(searchResult)
    assertEquals(expectedOutput, output)
  }

  test("test match length longer colorize search result #2") {
    val settings = SearchSettings(colorize = true, maxLineLength = 100)
    val pattern = "\\d+maxlen\\d+"
    val filePath = "./maxlen.txt"
    val lineNum = 10
    val matchStartIndex = 11
    val matchEndIndex = 120
    val line =
      "ABCDEFGHIJ0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789ABCDEFGHIJ"
    val searchResult = getSearchResult(pattern, filePath, FileType.Text, lineNum,
      matchStartIndex, matchEndIndex, Some(line), List.empty, List.empty)
    val expectedLine = "..." +
      ConsoleColor.GREEN.toString +
      "3456789012345678901234567890123456789012345678901maxlen890123456789012345678901234567890123456" +
      ConsoleColor.RESET.toString +
      "..."
    val expectedOutput = "%s: %d: [%d:%d]: %s".format(filePath,
      lineNum, matchStartIndex, matchEndIndex, expectedLine)
    val output = new SearchResultFormatter(settings).format(searchResult)
    assertEquals(expectedOutput, output)
  }

  test("test binary file search result") {
    val pattern = "Search"
    val filePath = "~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.exe"
    val lineNum = 0
    val matchStartIndex = 0
    val matchEndIndex = 0
    val line: Option[String] = None
    val searchResult = getSearchResult(pattern, filePath, FileType.Binary, lineNum,
      matchStartIndex, matchEndIndex, line)
    val expectedOutput = "%s matches at [%d:%d]".format(filePath, matchStartIndex, matchEndIndex)
    val output = new SearchResultFormatter(SearchSettings()).format(searchResult)
    assertEquals(expectedOutput, output)
  }

  test("test multi-line search result") {
    val pattern = "Search"
    val filePath = "~/src/git/xsearch/csharp/CsSearch/CsSearch/Searcher.cs"
    val lineNum = 10
    val matchStartIndex = 15
    val matchEndIndex = 23
    val line = "\tpublic class Searcher\n"
    val linesBefore = List("namespace CsSearch\n", "{\n")
    val linesAfter = List("\t{\n", "\t\tprivate readonly FileTypes _fileTypes;\n")
    val searchResult = getSearchResult(pattern, filePath, FileType.Code, lineNum,
      matchStartIndex, matchEndIndex, Some(line), linesBefore, linesAfter)
    val expectedOutput =
      """================================================================================
        |%s: %d: [%d:%d]
        |--------------------------------------------------------------------------------
        |   8 | namespace CsSearch
        |   9 | {
        |> 10 | 	public class Searcher
        |  11 | 	{
        |  12 | 		private readonly FileTypes _fileTypes;
        |""".stripMargin.format(filePath, lineNum, matchStartIndex, matchEndIndex)
    val output = new SearchResultFormatter(SearchSettings(colorize = false)).format(searchResult)
    assertEquals(expectedOutput, output)
  }
}

File: dart/dartsearch/test/search_result_test.dart
import 'dart:io' show File;

import 'package:dartfind/dartfind.dart';
import 'package:dartsearch/dartsearch.dart';
import 'package:test/test.dart';

void main() {
  test('test single-line search result', () {
    var settings = SearchSettings();
    settings.colorize = false;
    var formatter = SearchResultFormatter(settings);
    var pattern = RegExp('Search');
    var path = './lib/src/searcher.dart';
    var file = File(path);
    var fileResult = FileResult(file, FileType.code, 0, null);
    var lineNum = 13;
    var matchStartIndex = 7;
    var matchEndIndex = 15;
    var line = 'class Searcher {\n';
    var result = SearchResult(pattern, fileResult, lineNum, matchStartIndex,
        matchEndIndex, line, [], []);
    var expectedOutput =
        '$path: $lineNum: [$matchStartIndex:$matchEndIndex]: ${line.trim()}';
    var output = formatter.format(result);
    expect(output, expectedOutput);
  });

  test('test search result with line longer than maxlinelength', () {
    var settings = SearchSettings();
    settings.colorize = false;
    settings.maxLineLength = 100;
    var formatter = SearchResultFormatter(settings);
    var pattern = RegExp('maxlen');
    var path = './maxlen.txt';
    var file = File(path);
    var fileResult = FileResult(file, FileType.text, 0, null);
    var lineNum = 1;
    var matchStartIndex = 53;
    var matchEndIndex = 59;
    var line =
        '0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789';
    var result = SearchResult(pattern, fileResult, lineNum, matchStartIndex,
        matchEndIndex, line, [], []);
    var expectedLine =
        '...89012345678901234567890123456789012345678901maxlen89012345678901234567890123456789012345678901...';
    var expectedOutput =
        '$path: $lineNum: [$matchStartIndex:$matchEndIndex]: $expectedLine';
    var output = formatter.format(result);
    expect(output, expectedOutput);
  });

  test('test search result with colorized line longer than maxlinelength', () {
    var settings = SearchSettings();
    settings.colorize = true;
    settings.maxLineLength = 100;
    var formatter = SearchResultFormatter(settings);
    var pattern = RegExp('maxlen');
    var path = './maxlen.txt';
    var file = File(path);
    var fileResult = FileResult(file, FileType.text, 0, null);
    var lineNum = 1;
    var matchStartIndex = 53;
    var matchEndIndex = 59;
    var line =
        '0123456789012345678901234567890123456789012345678901maxlen8901234567890123456789012345678901234567890123456789';
    var result = SearchResult(pattern, fileResult, lineNum, matchStartIndex,
        matchEndIndex, line, [], []);
    var expectedLine =
        '...89012345678901234567890123456789012345678901${ConsoleColor.green}maxlen${ConsoleColor.reset}89012345678901234567890123456789012345678901...';
    var expectedOutput =
        '$path: $lineNum: [$matchStartIndex:$matchEndIndex]: $expectedLine';
    var output = formatter.format(result);
    expect(output, expectedOutput);
  });

  test('test binary file search result', () {
    var settings = SearchSettings();
    var formatter = SearchResultFormatter(settings);
    var pattern = RegExp('Search');
    var path = './lib/src/searcher.exe';
    var file = File(path);
    var fileResult = FileResult(file, FileType.binary, 0, null);
    var lineNum = 0;
    var matchStartIndex = 0;
    var matchEndIndex = 0;
    var result = SearchResult(pattern, fileResult, lineNum, matchStartIndex,
        matchEndIndex, null, [], []);
    var expectedOutput = '$path matches at [$matchStartIndex:$matchEndIndex]';
    var output = formatter.format(result);
    expect(output, expectedOutput);
  });

  test('test multiline search result', () {
    var settings = SearchSettings();
    settings.colorize = false;
    settings.linesBefore = 2;
    settings.linesAfter = 2;
    var formatter = SearchResultFormatter(settings);
    var pattern = RegExp('Searcher');
    var path = '~/src/xsearch/csharp/CsSearch/CsSearch/Searcher.cs';
    var file = File(path);
    var fileResult = FileResult(file, FileType.code, 0, null);
    var lineNum = 10;
    var matchStartIndex = 15;
    var matchEndIndex = 23;
    var line = '\tpublic class Searcher';
    var linesBefore = ['namespace CsSearch', '{'];
    var linesAfter = ['\t{', '\t\tprivate readonly FileTypes _fileTypes;'];
    var result = SearchResult(pattern, fileResult, lineNum, matchStartIndex,
        matchEndIndex, line, linesBefore, linesAfter);
    var expectedOutput =
        '================================================================================\n'
        '$path: $lineNum: [$matchStartIndex:$matchEndIndex]\n'
        '--------------------------------------------------------------------------------\n'
        '   8 | namespace CsSearch\n'
        '   9 | {\n'
        '> 10 | 	public class Searcher\n'
        '  11 | 	{\n'
        '  12 | 		private readonly FileTypes _fileTypes;\n';

    var output = formatter.format(result);
    expect(output, expectedOutput);
  });
}
